### ISSUE TITLE
Terraform plan not able to refresh

### ISSUE THREAD

**Issue by saikrishna547** [splunk-terraform/terraform-provider-signalfx#175]
The terraform plan is always getting struck , not throwing any errors and not even moving forward after certain steps.

Common thing observed is It always gets struck at data_links part 

<img width="791" alt="Screen Shot 2020-03-18 at 8 42 36 AM" src="https://user-images.githubusercontent.com/44275490/76980512-83298000-68f6-11ea-8ac9-eaefe319e0fe.png">

**Comment by cory-signalfx** [comment]
Hello, @saikrishna547. Can you provide a bit more information on this? It's possible it stems from some sort of rate limiting.

* Does signaling (INT or KILL) provide any feedback in the terminal or in a crash log?
* Does setting `-parellelism=1` stop the problem?
* Does it ever return? (I've experienced hanging from time to time, which is just a function of API requests being slow for whatever reason)

Thanks!

**Comment by saikrishna547** [comment]
I dont see any error when i kept if for 20-30+ mins.It's still in the hang state not seen a new line after that.
-parellelism=1 doesn't stop the problem that is the command i use in general. 

I dont see any issues with TF_DEBUG as well.
<img width="1086" alt="Screen Shot 2020-03-18 at 10 38 39 AM" src="https://user-images.githubusercontent.com/44275490/76990128-b2df8480-6904-11ea-9df1-043e020f73b2.png">

**Comment by cory-signalfx** [comment]
Weird! What version of Terraform are you using?

**Comment by saikrishna547** [comment]
@cory-signalfx  
Terraform v0.12.21
+ provider.aws v2.53.0
+ provider.signalfx v4.14.0

**Comment by cory-signalfx** [comment]
Thanks for sharing the `TF_LOG` output. As you said, it doesn't share much that helps us. I will try and reproduce on my end by creating/updating a lot of data links. I can't imagine why that would cause problems, but I've seen weirder things. :)

Any help you can provide with a triage is welcome. I'm sorry you keep hitting such interesting problems!

**Comment by cory-signalfx** [comment]
I've set up a local test that creates/updates/etc about 20 data links and can't reproduce this. I didn't think that was the problem, but I wanted to eliminate a simple rate limit problem for the data link endpoint.

**Comment by cory-signalfx** [comment]
If you can provide me a terraform source that repros this — which I realize you might not — that'd be super helpful. :)

**Comment by saikrishna547** [comment]
I observed few things today,I will share the info and let me know how we can debug this.
If i don't use parallelism=1, I do see all the data_links (16/16 ) got refreshed but plan is still in hang state.
if i use the parallelism=1 , I do see 10 or 11 data_links refreshed out of 16 and plan is still in hang state.


Let me know if you need any other details.

**Comment by saikrishna547** [comment]
Also, let me know what source code are you expecting?

Below is the DEBUG logs for that 
[tf2.log ](https://github.com/terraform-providers/terraform-provider-signalfx/files/4357915/tf2.log)

**Comment by saikrishna547** [comment]
Hope this might give you more clarity on the issue happening. Performed TG_LOG=TRACE and observed that signalfx provider is waiting for response of data_link resource which eventually blocked root to exit.
Please look at the below logs for information.

[tftrace.log ](https://github.com/terraform-providers/terraform-provider-signalfx/files/4358107/tftrace.log)

**Comment by cory-signalfx** [comment]
Thanks for these. It confirms what you're seeing, but I'm at a loss for why it would be blocked.

While trying to reproduce this I found a bug related to setting `is_default` and fixed it in #176. It would've affected you. I'll release a new version later today, but I don't think it's related to this problem.

**Comment by saikrishna547** [comment]
After trying multiple times, i can say that there is a problem only with data_links but not sure what exactly is the problem and thought you might know it better.
This is effecting my pipeline.
Let me know if you need anything from me.

**Comment by cory-signalfx** [comment]
Ok! I will check with engineering today and see if there's some sort of sneaky rate limiting or circular lock going on!

**Comment by saikrishna547** [comment]
Thank you :)

**Comment by cory-signalfx** [comment]
Ok, did some asking around and we can't think of anything that would cause this at a technical level.

Another q: You don't use any sort of a proxy to get to the SignalFx API, right?

**Comment by cory-signalfx** [comment]
Hey @saikrishna547. As is common in Go there were no timeouts in this transaction, so I've added them in #177. If the problem is in communication with the API as we suspect then we should see timeouts and errors which should help us track this down. I've asked Hashi for a build, which should be `v4.18.4`. At the moment they've not done the build yet, so we'll have to wait until that completes. If it does so before my end of day I'll update you. Otherwise you're on your own to check. :)

**Comment by saikrishna547** [comment]
Thanks for the update

**Comment by saikrishna547** [comment]
Another q: You don't use any sort of a proxy to get to the SignalFx API, right? 
No, we dont.

**Comment by cory-signalfx** [comment]
@saikrishna547 You should be able to update and try this again and get an error from a timeout!

**Comment by saikrishna547** [comment]
After how many minutes, do we get a timeout error? It's in hang state since 6 minutes and didn't get any error till now.

**Comment by cory-signalfx** [comment]
Seconds! 5 for connections and 30 for the entire transaction. Can you confirm you're running `v418.4` of the provider?

**Comment by saikrishna547** [comment]
Yes
<img width="868" alt="Screen Shot 2020-03-23 at 9 00 08 AM" src="https://user-images.githubusercontent.com/44275490/77336486-cd3ea700-6ce4-11ea-9944-e669360cdb95.png">

**Comment by cory-signalfx** [comment]
Wow, the plot thickens! That implies this isn't the HTTP stuff. That's positive to me cuz we looked and couldn't find any really slow transactions in our metrics. It's a negative now cuz wtf :)

So this leaves us with `dtruss`. Can you run `dtruss -p $PID` on the stuck process to see what it's doing?

**Comment by saikrishna547** [comment]
Hey cory, here you go.

[file1.zip](https://github.com/terraform-providers/terraform-provider-signalfx/files/4376328/file1.zip)
Is this what you are looking for?

**Comment by cory-signalfx** [comment]
Sorta! When the terraform run is hung I'm hoping to understand what the process is doing. If you run `dtruss -p $PID` where `$PID` is the PID of the terraform program then I'm hoping to see some output that helps us. If you have other ideas, I'm excited to hear them.

The output you gave me has nothing in it. Perhaps there are multiple PIDs and/or this needs `-parallelism 1`?

That aside, I believe we've eliminated the obvious HTTP/API problems with #177. Are you using a remote backend for state? That may also be a contributor… Without a way to reproduce this locally I'm out of ideas. :(

**Comment by saikrishna547** [comment]
Cory, I am using the remote backend for state. This is the logs for that 
when i ran dtruss where terraform -parallelism=1
[new.log.zip](https://github.com/terraform-providers/terraform-provider-signalfx/files/4379615/new.log.zip)
I will also try to change the backend to local machine and perform the above steps

**Comment by cory-signalfx** [comment]
That log is full of what we'd expect: waiting. 😄 Can't tell what it's waiting on though. :(

Let me know how your local test works!

**Comment by saikrishna547** [comment]
Hey @cory-signalfx  just performed terraform plan and apply for new environment and saw the similar issue reproduced even the state being local. It doesn't have anything other than signalfx and terraform in my code. ( I mean aws is removed now) . So, let me know what all you need from my end?

**Comment by cory-signalfx** [comment]
Ok, well, that's another possibility eliminated!

Since you've been able to pare this down to just SignalFx Terraform config, is it possible for you to share that with us? (If you prefer to do so non-publicly, you can send it to support!) It's ok if we don't have any of the metrics or anything, since this is about the dashboards and links.

**Comment by saikrishna547** [comment]
I will talk with support now and tell them to send it to you.

**Comment by cory-signalfx** [comment]
Ok! I'll keep an eye out and hopefully I can reproduce locally, thank you @saikrishna547!

**Comment by saikrishna547** [comment]
Also, not sure if this is also a bug. Can't we create 2 datalinks for same property pointing to same target dashboard keeping one default to true and the other one false? I can actually do this in the UI. But not in the terraform 
`resource "signalfx_data_link" "mydatalink_ec2_instances_2" {
  context_dashboard_id = "${signalfx_dashboard.ec2_instances.id}"
  property_name = "InstanceId"
  target_signalfx_dashboard {
    is_default = false
    name = "instance_id data link to ec2_instance dashboard"
    dashboard_group_id = "${signalfx_dashboard_group.mdcat_dashboard_group.id}"
    dashboard_id = "${signalfx_dashboard.ec2_instance1.id}"
  }
}

resource "signalfx_data_link" "mydatalink_ec2_instances_3" {
  context_dashboard_id = "${signalfx_dashboard.ec2_instances.id}"
  property_name = "InstanceId"
  target_signalfx_dashboard {
    is_default = true
    name = " instance_id data link to host dashboard"
    dashboard_group_id = "${signalfx_dashboard_group.mdcat_dashboard_group.id}"
    dashboard_id = "${signalfx_dashboard.host.id}"
  }
}`

**Comment by cory-signalfx** [comment]
Let's make that one a separate issue, I believe I've seen that as well.

**Comment by saikrishna547** [comment]
Yup sure. I guess you have received my repo from Katie. Let me know if you need anything from me.

**Comment by cory-signalfx** [comment]
I got it, thanks! Gimme a bit, as I've got a few other meetings today. I will report my findings!

**Comment by cory-signalfx** [comment]
Update from my end, I've received the configuration files and run apply and plan successfully. I didn't encounter any freezes or hangs. On your end is this hanging any time you do a `terraform plan`?

**Comment by saikrishna547** [comment]
Initially, when i do a plan I don't see any issue but after doing an apply and then try to add something new or delete something you should be able to reproduce this issue.
First apply then try to do a plan and let me know.

**Comment by cory-signalfx** [comment]
Hey @saikrishna547, I've just spent a while modifying, adding, and deleting data links and not been able to reproduce the problem. I also tried messing with dashboards and still nothing is causing a hang.

Can you give me any more thoughts on how I might reproduce?

**Comment by saikrishna547** [comment]
I have tried more than couple times deleted the exisiting state and then tried with new **terraform init** and this issue is able to reproduce for me even though.

**Comment by cory-signalfx** [comment]
Ok! I'll try a few more times through the day with more changes and see what I can do.

**Comment by cory-signalfx** [comment]
I've tried editing links and dashboards, adding new links, removing links, and combining all of the above and not been able to reproduce any sort of hang. I'll ask the in the Terraform committer's channel for a hand, but I'm out of ideas here. Is this something that only you are experiencing?

What is your OS, etc? Maybe this is happening on Linux or Windows but not on my Mac?

**Comment by cory-signalfx** [comment]
Asking around to the Hashi folks they suggested it might be related to [AWS credentials changing](https://www.reddit.com/r/Terraform/comments/bg7oe8/terraform_plan_hangs_without_making_any_attempts/) but you said you removed all the AWS so I'm still out of ideas. :(

**Comment by saikrishna547** [comment]
Hi @cory-signalfx  I never doubted my local machine since I run terraform for other projects which was working fine. But I can anyways run  this with a different machine and let you know. But, I have tried with my machine by removing existing state file and cloning new code still didn't worked for me.

**Comment by saikrishna547** [comment]
Hi @cory-signalfx Glad, it worked I  was able to do this with in Docker. Also, sometimes i face the timeout errors that you added while discussing in this issue. Is that something expected? Or any workarounds for you or me?

**Comment by cory-signalfx** [comment]
Glad you're able to confirm, but disappointed I still can't. :(

I have no other ideas for how to reproduce this locally and if I can't do that I'm unsure how to proceed. I believe we eliminated all the things that I or some contacts at HashiCorp could think of. 😢
